### PR TITLE
Feature/sg 1033 fix yolox anchors

### PR DESCRIPTION
**Pull request by BloodAxe**
[Ready for review]

Fixes Anchors implementation that prevents exporting a model to ONNX in FP16 mode

The biggest challenge was to ensure we still can load old checkpoints and new checkpoints (if training a model using current SG version) as well. 

# How

- Slightly refactored `adapt_state_dict_to_fit_model_layer_names` to remap layer weights to new names and extract all logic specific to YoloX to `YoloXCheckpointSolver`. 
- Removed old `_yolox_ckpt_solver` since it's logic now inside `YoloXCheckpointSolver`
 
# Notes

✅ This PR should be merged after https://github.com/Deci-AI/super-gradients/pull/1359 since it relies on changes from that PR

### CODE REVIEW

File: src/super_gradients/training/models/detection_models/yolo_base.py [hunk 1]
```diff
@@ -617,7 +619,7 @@ def _check_strides(self):
         if not torch.equal(m.stride, stride):
             raise RuntimeError("Provided anchor strides do not match the model strides")
 
-        self.register_buffer("stride", m.stride)  # USED ONLY FOR CONVERSION
+        # self.register_buffer("stride", m.stride)  # USED ONLY FOR CONVERSION
```
Comment (Louis-Dupont) [inline src/super_gradients/training/models/detection_models/yolo_base.py:622]: Leftover?

File: src/super_gradients/training/utils/checkpoint_utils.py [hunk 2]
```diff
@@ -44,6 +44,12 @@ def transfer_weights(model: nn.Module, model_state_dict: Mapping[str, Tensor]) -
             pass
 
 
+def maybe_remove_module_prefix(state_dict):
```
Comment (Louis-Dupont) [inline src/super_gradients/training/utils/checkpoint_utils.py:47]: I think it's quiet clear, but a simple docstring with an example is always better

File: src/super_gradients/training/utils/checkpoint_utils.py [hunk 3]
```diff
@@ -156,6 +161,1229 @@ def read_ckpt_state_dict(ckpt_path: str, device="cpu") -> Mapping[str, torch.Ten
         return state_dict
 
 
+class DefaultCheckpointSolver:
+    """
+    Default checkpoint solver class - has same behavior as PyTorch's default checkpoint solver.
+    """
```
Comment (Louis-Dupont) [inline src/super_gradients/training/utils/checkpoint_utils.py:167]: Would be great to have 2 words about this default behavior, not everyone would know what it means

File: src/super_gradients/training/utils/checkpoint_utils.py [hunk 4]
```diff
@@ -156,6 +161,1229 @@ def read_ckpt_state_dict(ckpt_path: str, device="cpu") -> Mapping[str, torch.Ten
...
+
+class YoloXCheckpointSolver:
+    @classmethod
+    def generate_mapping_table(cls) -> Mapping[str, str]:
+        """
+        Helper method to generate mapping table between olx YoloX checkpoints and the current YoloX layer names.
+        :return: A mapping dictionary {checkpoint_key: model_key}
+        """
+        from super_gradients.common.object_names import Models
+        from super_gradients.training import models
```
Comment (Louis-Dupont) [inline src/super_gradients/training/utils/checkpoint_utils.py:187]: Is there a motivation to not move these imports to the top of the module?

File: src/super_gradients/training/utils/checkpoint_utils.py [hunk 5]
```diff
@@ -156,6 +161,1229 @@ def read_ckpt_state_dict(ckpt_path: str, device="cpu") -> Mapping[str, torch.Ten
...
+                for k, v in checkpoint_state_dict.items()
+                if k not in {"stride", "_head.anchors._anchors", "_head.anchors._anchor_grid", "_head.anchors._stride", "_head._modules_list.14.stride"}
+            }
+
+            for (model_key, model_value), (checkpoint_key, checkpoint_value) in zip(model_state_dict.items(), new_sd.items()):
+                if model_value.size() == checkpoint_value.size() and model_key.split(".")[-1] == checkpoint_key.split(".")[-1]:
+                    if checkpoint_key in all_mapping_keys:
+                        assert all_mapping_keys[checkpoint_key] == model_key
+                    all_mapping_keys[checkpoint_key] = model_key
+                    print(f'"{checkpoint_key}": "{model_key}", ')
```
Comment (Louis-Dupont) [inline src/super_gradients/training/utils/checkpoint_utils.py:210]: Maybe `log` instead of `print`?

**Review by Louis-Dupont — COMMENTED** [review]
Some comments inline

File: src/super_gradients/training/utils/checkpoint_utils.py (hunk 4 above)
Comment (BloodAxe) [reply src/super_gradients/training/utils/checkpoint_utils.py:187]: Yes, this is to avoid circular imports.

File: src/super_gradients/training/utils/checkpoint_utils.py (hunk 3 above)
Comment (BloodAxe) [reply src/super_gradients/training/utils/checkpoint_utils.py:181]: Fixed in cc81568ae886751206245efe29ade055f7b4ce9f

Comment (BloodAxe) [reply src/super_gradients/training/utils/checkpoint_utils.py]: Fixed in a117c77942e0867765794b248ce02de132c65895

File: src/super_gradients/training/utils/checkpoint_utils.py (hunk 5 above)
Comment (BloodAxe) [reply src/super_gradients/training/utils/checkpoint_utils.py]: Fixed in 78f42e791c337bd10e8643e226864ba271e354ac

**Comment by BloodAxe** [comment]
Ran unit, integration - all greens

**Review by Louis-Dupont — APPROVED** [review]
LGTM, did you rerun the tests on tzags to make sure it works ?

**Comment by BloodAxe** [comment]
@Louis-Dupont yes, I re-run tests after we merged @shaydeci's PR with OD prediction once again
